### ISSUE TITLE
val within a parameterized (generic) inner class causes a NullPointerException upon compilation

### ISSUE THREAD

**Issue by lombokissues** [projectlombok/lombok#434]
_Migrated from Google Code (issue 361)_

**Comment by artgo** [comment]
Repro on Lombok 1.16.14 Oracle javac 1.8.0_112 Apache Maven 3.3.9
Compilation error:
```
[ERROR] Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException
```

```
Exception while resolving: NODE LOCAL (class com.sun.tools.javac.tree.JCTree$JCVariableDecl) @val()
final ___Lombok_VAL_Attrib__ newAppScreen = AppScreen.builder().pageContent(pageContent).build()
java.lang.NullPointerException
	at com.sun.tools.javac.comp.Check$Validator.visitTypeApply(Check.java:1259)
	at com.sun.tools.javac.tree.JCTree$JCTypeApply.accept(JCTree.java:2135)
	at com.sun.tools.javac.comp.Check$Validator.validateTree(Check.java:1359)
	at com.sun.tools.javac.comp.Check.validate(Check.java:1230)
	at com.sun.tools.javac.comp.Attr.visitVarDef(Attr.java:1055)
	at com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:852)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
	at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
	at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
	at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
	at lombok.javac.JavacResolution.attrib(JavacResolution.java:226)
	at lombok.javac.JavacResolution.memberEnterAndAttribute(JavacResolution.java:199)
	at lombok.javac.JavacResolution.resolveMethodMember(JavacResolution.java:146)
	at lombok.javac.handlers.HandleVal.visitLocal(HandleVal.java:138)
	at lombok.javac.JavacNode.traverse(JavacNode.java:100)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:143)
	at lombok.javac.JavacNode.traverse(JavacNode.java:85)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:143)
	at lombok.javac.JavacNode.traverse(JavacNode.java:75)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:143)
	at lombok.javac.JavacNode.traverse(JavacNode.java:70)
	at lombok.javac.JavacAST.traverse(JavacAST.java:139)
	at lombok.javac.HandlerLibrary.callASTVisitors(HandlerLibrary.java:257)
	at lombok.javac.JavacTransformer.transform(JavacTransformer.java:71)
	at lombok.javac.apt.LombokProcessor.process(LombokProcessor.java:250)
	at lombok.core.AnnotationProcessor$JavacDescriptor.process(AnnotationProcessor.java:114)
	at lombok.core.AnnotationProcessor.process(AnnotationProcessor.java:164)
	at lombok.launch.AnnotationProcessorHider$AnnotationProcessor.process(AnnotationProcessor.java:59)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
	at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:125)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:169)
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:825)
	at org.apache.maven.plugin.compiler.TestCompilerMojo.execute(TestCompilerMojo.java:153)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
```

**Comment by BullyWiiPlaza** [comment]
Can this please be fixed? I'm still getting the exception using `IntelliJ IDEA` and the [`Lombok` plugin](https://github.com/mplushnikov/lombok-intellij-plugin) v0.23-EAP on one of my projects:
`Error:java: Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException`

I don't know which piece of code causes it. Sometimes it goes through without an error though which is odd.

The `maven` dependency for `Lombok` is:

```
<dependency>
	<groupId>org.projectlombok</groupId>
	<artifactId>lombok</artifactId>
	<version>1.18.6</version>
	<scope>provided</scope>
</dependency>
```

**Comment by LaP0573** [comment]
Same here, getting the problem. We have a fairly large codebase in which we have spread Lombok (and `val`), and it gets very hard to pinpoint which vals should be removed to solve the issue. It's especially true as we only get that issue when compiling for Android and we spend most of our time testing on desktop, so sometimes it's weeks old code that will cause the problem.

**Comment by nhuanluong** [comment]
Lombok Plugin on IntelliJ has an issue with members has comments /* */. When members have a comment, it cannot build. 
example:
```
/* comment
* The issue of Lombok is here.
*/
DataTable datatable;
```

**Comment by nirujans** [comment]
I am also getting this issue with lombok v1.18.12

**Comment by BullyWiiPlaza** [comment]
**Update**
When building with maven, I actually get more details.

```
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ My-Project ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 324 source files to D:\Programs\Source Codes\Java\IntelliJ\My-Project\target\classes
Exception while resolving: NODE LOCAL (class com.sun.tools.javac.tree.JCTree$JCVariableDecl) @val()
final ___Lombok_VAL_Attrib__ iterationLocale = iterationLanguage.getLocale()(RegularFileObject[D:\Programs\Source Codes\Java\IntelliJ\My-Project\src\main\java\com\mycompany\translations\LanguageChooserManager.java])
java.lang.NullPointerException
        at com.sun.tools.javac.comp.Check$Validator.visitSelect(Check.java:1310)
        at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
        at com.sun.tools.javac.comp.Check$Validator.validateTree(Check.java:1359)
        at com.sun.tools.javac.comp.Check.validate(Check.java:1230)
        at com.sun.tools.javac.comp.Attr.visitVarDef(Attr.java:1055)
        at com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:852)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.visitForeachLoop(Attr.java:1168)
        at com.sun.tools.javac.tree.JCTree$JCEnhancedForLoop.accept(JCTree.java:1035)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
        at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.visitTry(Attr.java:1354)
        at com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1173)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
        at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.visitTry(Attr.java:1377)
        at com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1173)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
        at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.visitIf(Attr.java:1584)
        at com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1269)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
        at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
        at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
        at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
        at lombok.javac.JavacResolution.attrib(JavacResolution.java:239)
        at lombok.javac.JavacResolution.memberEnterAndAttribute(JavacResolution.java:207)
        at lombok.javac.JavacResolution.resolveMethodMember(JavacResolution.java:155)
        at lombok.javac.handlers.HandleVal.visitLocal(HandleVal.java:145)
        at lombok.javac.JavacNode.traverse(JavacNode.java:120)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:126)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:105)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:95)
        at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
        at lombok.javac.JavacNode.traverse(JavacNode.java:90)
        at lombok.javac.JavacAST.traverse(JavacAST.java:218)
        at lombok.javac.HandlerLibrary.callASTVisitors(HandlerLibrary.java:276)
        at lombok.javac.JavacTransformer.transform(JavacTransformer.java:64)
        at lombok.javac.apt.LombokProcessor.process(LombokProcessor.java:326)
        at lombok.core.AnnotationProcessor$JavacDescriptor.process(AnnotationProcessor.java:187)
        at lombok.core.AnnotationProcessor.process(AnnotationProcessor.java:237)
        at lombok.launch.AnnotationProcessorHider$AnnotationProcessor.process(AnnotationProcessor.java:90)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
        at com.sun.tools.javac.main.Main.compile(Main.java:523)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
        at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:126)
        at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:169)
        at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:785)
        at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:129)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:192)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.055 s
[INFO] Finished at: 2021-05-24T10:20:40+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project My-Project: Compilation failure
[ERROR] Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException
[ERROR]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```
I'm not even using an inner class. The piece of code is this:
```
// For caching
private static List<Language> languages;

// ...	

private void initializeResourceBundle()
{
	if (languageSelection != null
			&& languageSelection.getItemCount() > 0)
	{
		val language = languageSelection.getItemAt(languageSelection.getSelectedIndex());
		val locale = language.getLocale();
		setDefaultLocale(locale);
		resourceBundle = getBundle(resourceBundleName, locale);
		populateApplicationLanguages();
	} else
	{
		try
		{
			// Try the current language and country
			resourceBundle = getBundle(resourceBundleName);
		} catch (MissingResourceException exactBundleNotFound)
		{
			try
			{
				val locale = Locale.getDefault();
				val language = locale.getLanguage();

				// Try the same language only
				for (val iterationLanguage : languages) // <-- Changing this to "final Language" instead of val fixes the compilation issue
				{
					val iterationLocale = iterationLanguage.getLocale();
					val iterationLanguageTag = iterationLocale.getLanguage();

					if (iterationLanguageTag.equals(language))
					{
						setDefaultLocale(iterationLocale);
						resourceBundle = getBundle(resourceBundleName);

						return;
					}
				}

				throw new MissingResourceException("", "", "");
			} catch (MissingResourceException languageNotFound)
			{
				// Still error? Switch to the default language then
				resourceBundle = getBundle(resourceBundleName, defaultLocale);
				setDefaultLocale(defaultLocale);
			}
		}
	}

	if (resourceBundle == null)
	{
		// Should never happen, this is just an assertion
		throw new IllegalStateException("Resource bundle expected to be initialized!");
	}
}
```
`Language` is another class I defined in a different `.java` file. Strange.

Dependency:
```
<dependency>
	<groupId>org.projectlombok</groupId>
	<artifactId>lombok</artifactId>
	<version>1.18.18</version>
	<scope>provided</scope>
</dependency>
```

**Comment by rzwitserloot** [comment]
This may be very hard to fix and not worth fixing, given that instead we'd like to migrate `val` towards `var` instead. The 'fix' is then: Upgrade lombok (and be on at least java10).

**Comment by Rawi01** [comment]
@rzwitserloot I think I fixed this one in #2840, at least the stacktrace is similar
